### ISSUE TITLE
[Bug]: `occ db:convert-type` broken since `PrimaryReadReplicaConnection` support added

### ISSUE THREAD

**Issue by ksmonkey123** [nextcloud/server#45257]
### ⚠️ This issue respects the following points: ⚠️

- [X] This is a **bug**, not a question or a configuration/webserver/proxy issue.
- [X] This issue is **not** already reported on [Github](https://github.com/nextcloud/server/issues?q=is%3Aopen+is%3Aissue+label%3Abug) OR [Nextcloud Community Forum](https://help.nextcloud.com/) _(I've searched it)_.
- [X] Nextcloud Server **is** up to date. See [Maintenance and Release Schedule](https://github.com/nextcloud/server/wiki/Maintenance-and-Release-Schedule) for supported versions.
- [X] I agree to follow Nextcloud's [Code of Conduct](https://nextcloud.com/contribute/code-of-conduct/).

### Bug description

The OCC-Command `db:convert-type` seems to ignore the DB properties (type, user, host, database, port, password) provided through the CLI and instead operates on the database configured in `config.php`

I believe this may be related to the commit  79c4986 as this reworked the parameter handling in `ConnectionFactory.php`.

If I read this correctly, the default parameters are now always generated based on the system config instead of the passed arguments.

Running `db:convert-type --clean-schema` drops the tables in the "old" database.

### Steps to reproduce

1. Have Nextcloud 29.0.0 running with a MySQL database, and an empty PostgreSQL database
2. run `php occ db:convert-type --password=<PASSWORD> --port=5432 pgsql nextcloud_app postgres nextcloud` (note: the host `postgres` is a Kubernetes service name)
3. The command executes no matter if the Postgres Database is running or not, the command throws Constraint Violation errors in the table `oc_accounts`, I assume because it tries to migrate the content of the old database into itself.

### Expected behavior

`db:convert-type` 

### Installation method

Community Docker image

### Nextcloud Server version

29

### Operating system

Debian/Ubuntu

### PHP engine version

None

### Web server

None

### Database engine version

MySQL

### Is this bug present after an update or on a fresh install?

None

### Are you using the Nextcloud Server Encryption module?

None

### What user-backends are you using?

- [X] Default user-backend _(database)_
- [ ] LDAP/ Active Directory
- [ ] SSO - SAML
- [ ] Other

### Configuration report

```shell
{
    "system": {
        "htaccess.RewriteBase": "\/",
        "memcache.local": "\\OC\\Memcache\\APCu",
        "apps_paths": [
            {
                "path": "\/var\/www\/html\/apps",
                "url": "\/apps",
                "writable": false
            },
            {
                "path": "\/var\/www\/html\/custom_apps",
                "url": "\/custom_apps",
                "writable": true
            }
        ],
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "192.168.178.32",
            "cloud.awae.ch"
        ],
        "trusted_proxies": "***REMOVED SENSITIVE VALUE***",
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "dbtype": "mysql",
        "version": "29.0.0.19",
        "overwrite.cli.url": "https:\/\/cloud.awae.ch",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbport": "",
        "dbtableprefix": "oc_",
        "mysql.utf8mb4": true,
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true,
        "overwriteprotocol": "https",
        "theme": "",
        "loglevel": 2,
        "maintenance": false,
        "mail_smtpmode": "smtp",
        "mail_sendmailmode": "smtp",
        "mail_smtpauth": 1,
        "mail_smtpauthtype": "LOGIN",
        "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpport": "25",
        "mail_from_address": "***REMOVED SENSITIVE VALUE***",
        "mail_domain": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpname": "***REMOVED SENSITIVE VALUE***",
        "mail_smtppassword": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpsecure": "tls",
        "default_language": "de",
        "default_phone_region": "CH",
        "app_install_overwrite": [
            "polls"
        ]
    }
}
```

### List of activated Apps

```shell
Enabled:
  - activity: 2.21.1
  - circles: 29.0.0-dev
  - cloud_federation_api: 1.12.0
  - comments: 1.19.0
  - contactsinteraction: 1.10.0
  - dashboard: 7.9.0
  - dav: 1.30.1
  - federatedfilesharing: 1.19.0
  - federation: 1.19.0
  - files: 2.1.0
  - files_downloadlimit: 2.0.0
  - files_pdfviewer: 2.10.0
  - files_reminders: 1.2.0
  - files_sharing: 1.21.0
  - files_trashbin: 1.19.0
  - files_versions: 1.22.0
  - firstrunwizard: 2.18.0
  - logreader: 2.14.0
  - lookup_server_connector: 1.17.0
  - nextcloud_announcements: 1.18.0
  - notes: 4.10.0
  - notifications: 2.17.0
  - oauth2: 1.17.0
  - password_policy: 1.19.0
  - photos: 2.5.0
  - polls: 7.0.3
  - privacy: 1.13.0
  - provisioning_api: 1.19.0
  - recommendations: 2.1.0
  - related_resources: 1.4.0
  - serverinfo: 1.19.0
  - settings: 1.12.0
  - sharebymail: 1.19.0
  - support: 1.12.0
  - survey_client: 1.17.0
  - systemtags: 1.19.0
  - text: 3.10.0
  - theming: 2.4.0
  - twofactor_backupcodes: 1.18.0
  - updatenotification: 1.19.1
  - viewer: 2.3.0
  - workflowengine: 2.11.0
Disabled:
  - admin_audit: 1.19.0
  - bruteforcesettings: 2.9.0 (installed 2.4.0)
  - encryption: 2.17.0
  - extract: 1.3.6 (installed 1.3.6)
  - files_external: 1.21.0
  - files_rightclick: 0.15.1 (installed 1.6.0)
  - maps: 1.1.0 (installed 1.1.0)
  - suspicious_login: 7.0.0
  - twofactor_totp: 11.0.0-dev
  - user_ldap: 1.20.0
  - user_status: 1.9.0 (installed 1.4.0)
  - weather_status: 1.9.0 (installed 1.4.0)
```

### Nextcloud Signing status

```shell
No errors have been found.
```

### Additional info

* Docker Image: `nextcloud:29.0.0`
* MySQL Database: `mysql:8.4.0`
* Postgres Databse: `postgres:16.3`

**Comment by ksmonkey123** [comment]
Logging the Parameter `$additionalConnectionParams` in `ConnectionFactory.php` on [Line 166](https://github.com/nextcloud/server/blob/master/lib/private/DB/ConnectionFactory.php#L166) shows the following values for the "fromDB" and the "toDB":

# FromDB
```
Array
(
    [adapter] => OC\DB\AdapterMySQL
    [charset] => utf8mb4
    [driver] => pdo_mysql
    [wrapperClass] => OC\DB\Connection
    [driverOptions] => Array
        (
            [1005] => 1
        )

    [user] => nextcloud
    [password] => *************
    [host] => nextcloud-db
    [dbname] => nextcloud
    [tablePrefix] => oc_
    [sqlite.journal_mode] => WAL
    [defaultTableOptions] => Array
        (
            [collate] => utf8mb4_bin
            [charset] => utf8mb4
            [tablePrefix] => oc_
        )

    [primary] => Array
        (
            [adapter] => OC\DB\AdapterMySQL
            [charset] => utf8mb4
            [driver] => pdo_mysql
            [wrapperClass] => OC\DB\Connection
            [driverOptions] => Array
                (
                    [1005] => 1
                )

            [user] => nextcloud
            [password] => ****************
            [host] => nextcloud-db
            [dbname] => nextcloud
            [tablePrefix] => oc_
            [sqlite.journal_mode] => WAL
            [defaultTableOptions] => Array
                (
                    [collate] => utf8mb4_bin
                    [charset] => utf8mb4
                    [tablePrefix] => oc_
                )

        )

    [replica] => Array
        (
            [0] => Array
                (
                    [adapter] => OC\DB\AdapterMySQL
                    [charset] => utf8mb4
                    [driver] => pdo_mysql
                    [wrapperClass] => OC\DB\Connection
                    [driverOptions] => Array
                        (
                            [1005] => 1
                        )

                    [user] => nextcloud
                    [password] => *************
                    [host] => nextcloud-db
                    [dbname] => nextcloud
                    [tablePrefix] => oc_
                    [sqlite.journal_mode] => WAL
                    [defaultTableOptions] => Array
                        (
                            [collate] => utf8mb4_bin
                            [charset] => utf8mb4
                            [tablePrefix] => oc_
                        )

                )

        )

)

```

# ToDB
```
Array
(
    [adapter] => OC\DB\AdapterMySQL
    [charset] => utf8mb4
    [driver] => pdo_mysql
    [wrapperClass] => OC\DB\Connection
    [driverOptions] => Array
        (
            [1005] => 1
        )

    [user] => nextcloud_app
    [password] => ***************
    [host] => postgres
    [dbname] => nextcloud
    [tablePrefix] => oc_
    [sqlite.journal_mode] => WAL
    [defaultTableOptions] => Array
        (
            [collate] => utf8mb4_bin
            [charset] => utf8mb4
            [tablePrefix] => oc_
        )

    [primary] => Array
        (
            [adapter] => OC\DB\AdapterMySQL
            [charset] => utf8mb4
            [driver] => pdo_mysql
            [wrapperClass] => OC\DB\Connection
            [driverOptions] => Array
                (
                    [1005] => 1
                )

            [user] => nextcloud
            [password] => *************
            [host] => nextcloud-db
            [dbname] => nextcloud
            [tablePrefix] => oc_
            [sqlite.journal_mode] => WAL
            [defaultTableOptions] => Array
                (
                    [collate] => utf8mb4_bin
                    [charset] => utf8mb4
                    [tablePrefix] => oc_
                )

        )

    [replica] => Array
        (
            [0] => Array
                (
                    [adapter] => OC\DB\AdapterMySQL
                    [charset] => utf8mb4
                    [driver] => pdo_mysql
                    [wrapperClass] => OC\DB\Connection
                    [driverOptions] => Array
                        (
                            [1005] => 1
                        )

                    [user] => nextcloud
                    [password] => **********
                    [host] => nextcloud-db
                    [dbname] => nextcloud
                    [tablePrefix] => oc_
                    [sqlite.journal_mode] => WAL
                    [defaultTableOptions] => Array
                        (
                            [collate] => utf8mb4_bin
                            [charset] => utf8mb4
                            [tablePrefix] => oc_
                        )

                )

        )

    [port] => 5432
)
```

It is notable that in the second case (which should be the postgres-database configured via the CLI arguments), the host, dbname and port in the top-level array are set correctly from the CLI, Both the `primary` and the `replica` fields and even some top-level fields obviously use the DB-configs from `config.php`

**Comment by joshtrichards** [comment]
Possibly related: #45097

**Comment by intersectRaven** [comment]
I can verify but can't contribute logs. I used the --clear-schema option in my attempt and my original MariaDB DB got deleted. Good thing I have backups.

**Comment by Delagen** [comment]
I modified https://github.com/nextcloud/server/blob/master/core/Command/Db/ConvertType.php#L247
to
```
$connectionParams = $this->connectionFactory->getDefaultConnectionParams($type);
```
and seems all converted

Also have to remove mysql.utf8mb4 parameter from config.php to perform conversion to remove client_encoding from params of DbFactory

**Comment by HammyHavoc** [comment]
Got hit by this a week ago. Is that particular Nextcloud instance completely screwed?

**Comment by joshtrichards** [comment]
Yes, unfortunately. :cry:  The end result is the parameters get swapped so it likely dropped the source database.

Fortunately, recovery should be a simple matter of restoring from backup (just the db portion): https://docs.nextcloud.com/server/latest/admin_manual/maintenance/restore.html#restore-database

**Comment by joshtrichards** [comment]
@Delagen Mind pushing that as PR? I know there may be additional changes needed, but it'll start things in motion.

**Comment by Delagen** [comment]
@joshtrichards I don't mind that my changes solve the problem correctly. It only makes possible for me successful conversion from MySQL to Postgres

**Comment by chimpboy** [comment]
I have just experienced this problem on a new installation.

**Comment by rotdrop** [comment]
Would not be it the simplest solution, if the ConnectionFactory::getConnection() method would first check if the `$additionalConnectionParams` argument contains explicit connection parameters and in that case simply skip that fancy primary / replica stuff? At least for the case were the explicitly specified connection differs from the "internal" database connection configured in `config/config.php`. In principle this should restore the previous behaviour when using the factory for alternate database connections.

**Comment by szaimen** [comment]
Hi, is this https://github.com/nextcloud/server/pull/45013 fixing the problem?

**Comment by jiriks74** [comment]
I just looked into it a bit and so far it looks like the databases get mixed up somewhere in the script.

When I hardcoded the `toDB` values it tried to connect to the target database (PostgreSQL) with the credentials of my current database (MariaDB/MySQL). It used the username from the old database on the new one.

**Comment by juliushaertl** [comment]
Possible patch candidate, but haven't had time to verify and test this yet:

```diff
diff --git a/core/Command/Db/ConvertType.php b/core/Command/Db/ConvertType.php
index 333f29625f6..078fd47965b 100644
--- a/core/Command/Db/ConvertType.php
+++ b/core/Command/Db/ConvertType.php
@@ -227,15 +227,22 @@ class ConvertType extends Command implements CompletionAwareInterface {
        protected function getToDBConnection(InputInterface $input, OutputInterface $output) {
                $type = $input->getArgument('type');
                $connectionParams = $this->connectionFactory->createConnectionParams();
-               $connectionParams = array_merge($connectionParams, [
+
+               $overrideConnectionParams = [
                        'host' => $input->getArgument('hostname'),
                        'user' => $input->getArgument('username'),
                        'password' => $input->getOption('password'),
                        'dbname' => $input->getArgument('database'),
-               ]);
+               ];
                if ($input->getOption('port')) {
-                       $connectionParams['port'] = $input->getOption('port');
+                       $overrideConnectionParams['port'] = $input->getOption('port');
                }
+
+               $connectionParams = array_merge($connectionParams, [
+                       'primary' => $overrideConnectionParams,
+                       'replica' => [$overrideConnectionParams],
+               ]);
+
                return $this->connectionFactory->getConnection($type, $connectionParams);
        }

```

**Comment by szaimen** [comment]
Hi @intersectRaven @ksmonkey123 @Delagen @HammyHavoc @chimpboy @rotdrop are you able to test the above patch? Thanks in advance! :)

**Comment by jiriks74** [comment]
@szaimen I'll be probably playing around with a new (old) server this evening. I'll back up my database and try the patch out. I'll know practically immediately as I get the duplicate entry error right at the start if it tries to write into the original DB.

**Comment by jiriks74** [comment]
I tried it out, it doesn't work. I get stuck on `Creating schema in new database` even after hardcoding the connections for my databases to be sure that it's passed right. I'll try to find out where the issue is.

**Comment by jiriks74** [comment]
The conversion script gets stuck on this line

https://github.com/nextcloud/server/blob/34f5cba5b3cf84d5f85e95985b044ef9ab002f82/core/Command/Db/ConvertType.php#L210

**Comment by leon1995** [comment]
Are there any new information on this?

**Comment by zackgomez** [comment]
Any progress?

Is the current guidance to create a fresh system losing an existing data?

**Comment by Infamousbugg** [comment]
I also have a migration on hold waiting for this issue to be resolved.

**Comment by CorneliousJD** [comment]
Unfortunately continuing to wait. I continue to have DB corruption w/ MariaDB and am anxious to switch over to Postgres but this seemingly keeps getting pushed back?

**Comment by Investigamer** [comment]
If this issue is still weeks/months away from a proper fix, could we at least stem the bleeding with a FAQ or Wiki entry on how to manually migrate from one database format to another without using occ, citing this current issue and giving notice that a fix is forthcoming?

**Comment by rankerson** [comment]
> If this issue is still weeks/months away from a proper fix, could we at least stem the bleeding with a FAQ or Wiki entry on how to manually migrate from one database format to another without using occ, citing this current issue and giving notice that a fix is forthcoming?

I think it is a good sign, that there has been some action. I assume it is now on the direct todo list of the developers. So let us continue to hope :-)

**Comment by lwenderoth** [comment]
@Delagen wrote [here](https://github.com/nextcloud/server/issues/45257#issuecomment-2110831104) :
> I modified https://github.com/nextcloud/server/blob/master/core/Command/Db/ConvertType.php#L247 to
> 
> ```
> $connectionParams = $this->connectionFactory->getDefaultConnectionParams($type);
> ```
> 
> and seems all converted
> 
> Also have to remove mysql.utf8mb4 parameter from config.php to perform conversion to remove client_encoding from params of DbFactory

Did someone test this? It is now line 229 though.
I came up with the same solution.
The problem seems to be that createConnectionParams() always returns the currently configured DBs values while getDefaultConnectionParams($type) does what it's name suggests: Return default values for $type, NOT values for the configured DB.

**Comment by jiriks74** [comment]
It seems like a fix is finally on it's way (https://github.com/nextcloud/server/pull/46931) and it's been approved to be backported to v29 (https://github.com/nextcloud/server/pull/47046)

**Comment by joshtrichards** [comment]
Now that #47046 is merged in stable29 (you'll need to manually patch until officially released), if you wish to test a db conversion, you will also need to *revert* the code change made in #45469 (which disabled the `db:convert-type` command to avoid data loss).

I have not had a chance to test #47046 so I've not submitted a PR to revert the disabling of the command yet. (P.S. This doesn't have to wait on me).

**Comment by jiriks74** [comment]
I could get to it next week. I'm on holiday until Wednesday and I don't feel like converting remotely. If no one else gets to it I probably will by the end of the next week.

**Comment by TheTimeWalker** [comment]
I have tried the patch on my side, but it seems like that it still doesn't properly pass the sql type parameter. At least if you try to use something like mysql -> mysql or mysql -> sqlite3, occ properly rejects it. However, when you use `pgsql` then the connection param still defaults to pdo_mysql:

```json5
{
    "adapter": "OC\\DB\\AdapterMySQL",
    "charset": "utf8mb4",
    "driver": "pdo_mysql",
    "wrapperClass": "OC\\DB\\Connection",
    "driverOptions": {
        "1005": true
    },
    "user": "ncadmin",
    "password": "test",
    "host": "cloud-pg-temp",
    "dbname": "nextcloud_db",
    "tablePrefix": "oc_",
    "sqlite.journal_mode": "WAL",
    "defaultTableOptions": {
        "collate": "utf8mb4_bin",
        "charset": "utf8mb4",
        "tablePrefix": "oc_"
    },
...
```

**Comment by asmoker** [comment]
Any progress? 🤨

**Comment by Queuecumber** [comment]
> I have tried the patch on my side, but it seems like that it still doesn't properly pass the sql type parameter.

I can confirm this is still bugged even with the patch, the code as written doesn't initialize the "to" database connection correctly.

It looks like it's because https://github.com/nextcloud/server/blob/307608b26cc0492eda5657bf674ec634523a63ee/core/Command/Db/ConvertType.php#L229 calls the connection factory function `createConnectionParams();` which populates a bunch of defaults from the "from" database configuration which in my case is mysql. 

That comes from a change from 8 months ago:  79c4986354da2c7dd102de21174114e7687daf98 which adds the call to

```
$connectionParams = array_merge($this->getDefaultConnectionParams($type),  ...
```

namely `getDefaultConnectionParams` was not there before. So I think this has been broken for some time now.  I haven't been able to find a workaround of it yet.

CC @juliushaertl the author of the commit in question

**Comment by Queuecumber** [comment]
I finally got this running but it's quite hacky, I can post what I did if theres interest but there should be a proper solution to this

**Comment by normen** [comment]
> I finally got this running but it's quite hacky, I can post what I did if theres interest but there should be a proper solution to this

Yes please, how did you get it to work?

**Comment by Queuecumber** [comment]
Unfortunately after upgrading to .0.5 my changes are gone but I can try to recreate it although it may require some debugging to get it right

In `lib/private/DB/ConnectionFactory.php` I changed the function `createConnectionParams` to take a `type` parameter, so something like

```
public function createConnectionParams(string $configPrefix = '', array $additionalConnectionParams = [], $type = '') {
                if ($type === '') {
		        $type = $this->config->getValue('dbtype', 'sqlite');
                 }
``` 

Then I changed `core/Command/Db/ConvertType.php` to remove the guard and then in `getToDBConnection` I pass the type parameter to the updated function from above, should be something like:

```
$connectionParams = $this->connectionFactory->createConnectionParams($type);
```

I was converting from MariaDB to PG and I also have to set `mysql.utf8mb4 = false` in `config.php` for some reason, but after these updates it worked.

**Comment by normen** [comment]
Great, thanks, that works! For reference, the only line that I had to implement differently from the above was this (compatible to PHP 7):

`$connectionParams = $this->connectionFactory->createConnectionParams('', [], $type);`

**Comment by TjenWellens** [comment]
> Unfortunately after upgrading to .0.5 my changes are gone but I can try to recreate it although it may require some debugging to get it right
> 
> In `lib/private/DB/ConnectionFactory.php` I changed the function `createConnectionParams` to take a `type` parameter, so something like
> 
> ```
> public function createConnectionParams(string $configPrefix = '', array $additionalConnectionParams = [], $type = '') {
>                 if ($type === '') {
> 		        $type = $this->config->getValue('dbtype', 'sqlite');
>                  }
> ```
> 
> Then I changed `core/Command/Db/ConvertType.php` to remove the guard and then in `getToDBConnection` I pass the type parameter to the updated function from above, should be something like:
> 
> ```
> $connectionParams = $this->connectionFactory->createConnectionParams($type);
> ```
> 
> I was converting from MariaDB to PG and I also have to set `mysql.utf8mb4 = false` in `config.php` for some reason, but after these updates it worked.

For ppl who don't know what "remove the guard" refers to. Comment out the entire throw block like so.
```php
// WARNING:
// Leave in place until #45257 is addressed to prevent data loss (hopefully in time for th...
//throw new \InvalidArgumentException(
//		'This command is temporarily disabled (until the next maintenance release).'
//);
```

**Comment by Investigamer** [comment]
Did we really shift this to milestone 31...? After people were instructed to wait "until the next maintenance release" for several months? 😅

**Comment by lightonflux** [comment]
> I was converting from MariaDB to PG and I also have to set mysql.utf8mb4 = false in config.php for some reason, but after these updates it worked.

Do not do this, as this might break the source DB. Instead add a check for mysql in lib/private/DB/ConnectionFactory.php

```patch
--- lib/private/DB/ConnectionFactory.php.backup 2024-09-12 12:29:28.000000000 +0000
+++ lib/private/DB/ConnectionFactory.php        2024-09-15 17:57:06.051623401 +0000
@@ -120,7 +120,7 @@
                $normalizedType = $this->normalizeType($type);
                $eventManager = new EventManager();
                $eventManager->addEventSubscriber(new SetTransactionIsolationLevel());
-               $connectionParams = $this->createConnectionParams('', $additionalConnectionParams);
+               $connectionParams = $this->createConnectionParams('', $additionalConnectionParams, $type);
                switch ($normalizedType) {
                        case 'pgsql':
                                // pg_connect used by Doctrine DBAL does not support URI notation (enclosed in brackets)
@@ -191,8 +191,12 @@
         * @param string $configPrefix
         * @return array
         */
-       public function createConnectionParams(string $configPrefix = '', array $additionalConnectionParams = []) {
-               $type = $this->config->getValue('dbtype', 'sqlite');
+       // public function createConnectionParams(string $configPrefix = '', array $additionalConnectionParams = []) {
+       public function createConnectionParams(string $configPrefix = '', array $additionalConnectionParams = [], $type = '') {
+               if ($type === '') {
+                       $type = $this->config->getValue('dbtype', 'sqlite');
+               }
+               // $type = $this->config->getValue('dbtype', 'sqlite');
 
                $connectionParams = array_merge($this->getDefaultConnectionParams($type), [
                        'user' => $this->config->getValue($configPrefix . 'dbuser', $this->config->getValue('dbuser', '')),
@@ -224,7 +228,7 @@
                        'tablePrefix' => $connectionParams['tablePrefix']
                ];
 
-               if ($this->config->getValue('mysql.utf8mb4', false)) {
+               if ($this->config->getValue('mysql.utf8mb4', false) && $type==="mysql") {
                        $connectionParams['defaultTableOptions'] = [
                                'collate' => 'utf8mb4_bin',
                                'charset' => 'utf8mb4',

```

**Comment by joshtrichards** [comment]
> Did we really shift this to milestone 31...? After people were instructed to wait "until the next maintenance release" for several months? 😅

It'll be a bug fix. Any fix will be backported to whatever earlier majors need it. It can't have a v30 milestone since it's already released. Development occurs against the `master` branch (upcoming v31 at the moment). Bug fixes get backported after they're merged there (and then land in maintenance releases for earlier majors).

> I realized the developers of Nextcloud owe me absolutely nothing, but this has still been a frustrating experience.

I'm sorry it has been a frustrating experience. I'm hopeful someone more familiar that area of code than me (or with more time on their hands to get familiar with it) will take a closer took soon.

If you are a customer of Nextcloud GmbH, you can tell your contacts there through the appropriate channels that this Issue is a priority for you as a customer and ask them to fix it on your behalf. Otherwise you're at the mercy of whenever someone gets around to it in the project.

@CorneliousJD I'm curious about your corruption situation, but this isn't the place to follow-up on it. Have you posted about your corruption issues somewhere (such as on the community help forum) where I can take a look and gain some further context?

**Comment by CorneliousJD** [comment]
> @CorneliousJD I'm curious about your corruption situation, but this isn't the place to follow-up on it. Have you posted about your corruption issues somewhere (such as on the community help forum) where I can take a look and gain some further context?

I have not posted elsewhere about it in great detail, but it will randomly have corrupt tables and it only happens within the nextcloud database. I had previously run w/ postgres w/out these issues so I wanted to revert back to postgres and see how things go before abandoning Nextcloud, but setting up a whole new instance w/out doing a db convert just isn't an option for me right now.  

Before I dive too deep into it I am just waiting for the Postgres conversion to be back in the cards so I Can test that and see how it goes with that setup.

**Comment by Delogon** [comment]
For me the combined solutions of @Queuecumber, @normen and @lightonflux worked, and I was able to convert my mariadb to postgresql. 

So combined it would look like this:

In `lib/private/DB/ConnectionFactory.php`:
```diff
-               $connectionParams = $this->createConnectionParams('', $additionalConnectionParams);
+               $connectionParams = $this->createConnectionParams('', $additionalConnectionParams, $type);
```

```diff
-       public function createConnectionParams(string $configPrefix = '', array $additionalConnectionParams = []) {
-               $type = $this->config->getValue('dbtype', 'sqlite');
+       public function createConnectionParams(string $configPrefix = '', array $additionalConnectionParams = [], $type = '') {
+               if ($type === '') {
+                       $type = $this->config->getValue('dbtype', 'sqlite');
+               }
```

```diff
+               if ($type === '') {
+                       $type = $this->config->getValue('dbtype', 'sqlite');
+               }
-               if ($this->config->getValue('mysql.utf8mb4', false)) {
+               if ($this->config->getValue('mysql.utf8mb4', false) && $type==="mysql") {
```


In `core/Command/Db/ConvertType.php`
```diff
- $connectionParams = $this->connectionFactory->createConnectionParams();
+ $connectionParams = $this->connectionFactory->createConnectionParams('', [], $type);
```

Maybe someone else could also check this?

**Comment by cburlacu** [comment]
@Delogon It worked for me (v30). Finally I was able to migrate to pgsql. At the first glance, all seems to be OK. Thanks!

**Comment by Investigamer** [comment]
@Delogon this is the way, worked great sqlite -> psql. Cheers and good luck everyone.

**Comment by cektom** [comment]
@Delogon Thanks for the summary! Looks good on my end. Migrated from sqlite to mysql (mariadb  Ver 15.1 Distrib 10.6.18-MariaDB) with the latest docker image (Nextcloud 30.0.0)

**Comment by KoMa1012** [comment]
Still getting the error message:
```
In ConvertType.php line 158:

  This command is temporarily disabled (until the next maintenance release).
```
with command `php occ db:convert-type --all-apps --password mypassword pgsql ncadmin postgresql17 nextcloud_db` Running 30.0.0
![Image](https://github.com/user-attachments/assets/fbcc3f20-e5dc-4646-bc37-9acc30b784b6)

I'm running mariadb and want to switch to psql.

Any ideas if I'm doing something wrong or if this case has not been implemented yet?

**Comment by leon1995** [comment]
@KoMa1012 the issue has not been fixed officially. you can try the workaround provided by @Delogon. 
regarding the `whiteboard.svg` checkout https://github.com/nextcloud/server/issues/47985

**Comment by KoMa1012** [comment]
@Delogon I'm getting an error message:
```
In TableDoesNotExist.php line 16:

  There is no table with name "ncadmin.oc_group_folders_acl" in the schema.
```

I'm not sure if relevant, but this table is empty for me.

**Comment by WhiskyDelta** [comment]
@delogon Since your fix seems to work, do you want to make a pull request? If not, I might be able to do it.

**Comment by hagak** [comment]
So I am also trying to move from mariadb to postgres and on nextcloud v30.  I did the fix that @Delogon posted and that got me further.  However i end up with this during the conversion.

```
- oc_properties
 10/10 [============================] 100% < 1 sec/< 1 sec
In ExceptionConverter.php line 87:

  An exception occurred while executing a query: SQLSTATE[22021]: Character not in repertoire: 7 ERROR:  invalid byte sequence for encoding "UTF8": 0x00
  CONTEXT:  unnamed portal parameter $5


In Exception.php line 28:

  SQLSTATE[22021]: Character not in repertoire: 7 ERROR:  invalid byte sequence for encoding "UTF8": 0x00
  CONTEXT:  unnamed portal parameter $5


In Statement.php line 130:

  SQLSTATE[22021]: Character not in repertoire: 7 ERROR:  invalid byte sequence for encoding "UTF8": 0x00
  CONTEXT:  unnamed portal parameter $5
```
Any ideas?

**Comment by lightonflux** [comment]
@hagak what is your encoding in mysql?

**Comment by hagak** [comment]
> [@hagak](https://github.com/hagak) what is your encoding in mysql?

UTF8  the issue from what i gather in that other related bug that i apparently forgot to put the link to.  Shows that it is due to postgres schema using a TEXT column instead of a Bytea .  I believe there is a PR to resolve that as well

**Comment by Delogon** [comment]
I created the pull request #48666 but tbh I don't know if I did that correct, since creation nothing has happened. @joshtrichards or @juliusknorr maybe you could tell me what to do, to get it merged.

**Comment by daks** [comment]
I want to migration from a standalone nginx/php-fpm/mariadb Nextcloud installation to AIO "The official Nextcloud installation method"  but I'm blocked by this issue since several weeks.

Is the patch provided by @Delogon here https://github.com/nextcloud/server/issues/45257#issuecomment-2363167396 or in the PR #48666 safe to use?
Is there any chance to see the issue solved in the following releases?

**Comment by jiriks74** [comment]
> Is the patch provided by @Delogon here https://github.com/nextcloud/server/issues/45257#issuecomment-2363167396 or in the PR https://github.com/nextcloud/server/pull/48666 safe to use?

I cannot answer that but even if it is you should make a database backup anyway

**Comment by daks** [comment]
> > Is the patch provided by [@Delogon](https://github.com/Delogon) here [#45257 (comment)](https://github.com/nextcloud/server/issues/45257#issuecomment-2363167396) or in the PR [#48666](https://github.com/nextcloud/server/pull/48666) safe to use?
> 
> I cannot answer that but even if it is you should make a database backup anyway

Of course. I'm just worry if I use it, don't see any problem, start using my new postgres DB without detecting any data loss :-/

**Comment by thespielplatz** [comment]
> Is the patch provided by [@Delogon](https://github.com/Delogon) here [#45257 (comment)](https://github.com/nextcloud/server/issues/45257#issuecomment-2363167396) or in the PR [#48666](https://github.com/nextcloud/server/pull/48666) safe to use? Is there any chance to see the issue solved in the following releases?

I just applied the PR [#48666] manually on nextcloud [29.0.9-fpm-alpine](https://github.com/nextcloud/docker/blob/4f130dc7527e9f405349767596b168c81b45ba24/29/fpm-alpine/Dockerfile) and successfully converted from mysql to postgres. So far everyhting looks good.

**Comment by jiriks74** [comment]
I'll be, probably, trying to migrate this evening. I'll report how it goes.

**Comment by jiriks74** [comment]
Well it failed while creating schema for me:
![Image](https://github.com/user-attachments/assets/c1f36ac0-c17f-44a1-8ace-bdff49f22801)

But that's probably some other issue.

**Comment by Delogon** [comment]
> Well it failed while creating schema for me: ![Image](https://github.com/user-attachments/assets/c1f36ac0-c17f-44a1-8ace-bdff49f22801)
> 
> But that's probably some other issue.

That bug is already known. A potential solution would be to disable the group folder plugin, but I do not know how suitable that is for you.

> [@Delogon](https://github.com/Delogon) I'm getting an error message:
> 
> ```
> In TableDoesNotExist.php line 16:
> 
>   There is no table with name "ncadmin.oc_group_folders_acl" in the schema.
> ```
> 
> I'm not sure if relevant, but this table is empty for me. SQL Commands to generate this table in mariadb:
> 
> ```
> SET SQL_MODE = "NO_AUTO_VALUE_ON_ZERO";
> START TRANSACTION;
> SET time_zone = "+00:00";
> 
> CREATE TABLE `oc_group_folders_acl` (
>   `acl_id` bigint(20) NOT NULL,
>   `fileid` bigint(20) NOT NULL,
>   `mapping_type` varchar(16) NOT NULL,
>   `mapping_id` varchar(64) NOT NULL,
>   `mask` smallint(6) NOT NULL,
>   `permissions` smallint(6) NOT NULL
> ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;
> ALTER TABLE `oc_group_folders_acl`
>   ADD PRIMARY KEY (`acl_id`),
>   ADD UNIQUE KEY `groups_folder_acl_unique` (`fileid`,`mapping_type`,`mapping_id`),
>   ADD KEY `groups_folder_acl_mapping` (`mapping_type`,`mapping_id`);
> ALTER TABLE `oc_group_folders_acl`
>   MODIFY `acl_id` bigint(20) NOT NULL AUTO_INCREMENT;
> COMMIT;
> ```
> 
> removing the group folder plugin solves the issue... however, now the group folders are gone, but this I can survive.

**Comment by jiriks74** [comment]
> That bug is already known. A potential solution would be to disable the group folder plugin, but I do not know how suitable that is for you.

That crossed my mind but Nextcloud kinda doesn't want to do it. I don't use the plugin but it was disabled but installed.

While writing this I managed to removed it with `occ app:remove` and I started the conversion.

**Comment by jiriks74** [comment]
Ok. The database migration itself went smoothly after disabling group folders. Importing data into AIO is the part that was full of pain.

It had issues with the database dump, then the data files, then 3 apps that I had installed, ... But it seems to be done now.

MAKE BACKUPS (they saved me)

**Comment by daks** [comment]
disabling user theming seems sufficient

**Comment by dark-vex** [comment]
I've applied the patch too and tried to run the migration, got stuck with the same issue of @daks. After fixing it I was able to perform the migration, however I've got these two tables not being converted:
```
The following tables will not be converted:
oc_file_metadata
oc_twofactor_u2f_registrations
Continue with the conversion (y/n)? [n] y
```

For `oc_file_metadata` should not be an issue since I should be able to re-generate it with `occ files:scan --all --generate-metadata`. Regarding `oc_twofactor_u2f_registrations`, I don't have 2FA currently set so in my case it's not a problem but I guess it would be for whom have it enabled.

**Comment by dark-vex** [comment]
> > Regarding `oc_twofactor_u2f_registrations`, I don't have 2FA currently set so in my case it's not a problem but I guess it would be for whom have it enabled.
> 
> This is weird as I have 2FA enabled and I set up. Migration worked without issues in regards to 2FA as it's still working.

I think that probably it does print that message because the table is empty and there is nothing to migrate:
```
MariaDB [nextcloud]> select * from oc_twofactor_u2f_registrations;
Empty set (0.000 sec)

MariaDB [nextcloud]>
```

**Comment by le-patenteux** [comment]
I just discovered that the db:convert-type has been broken for a while, as I was needing it! 
Is there a way to go from mariaDB to PostgreSQL for Nextcloud, without using occ ? Is there a command in the wild that would migrate the DB in a nextcloud compatible way?

Regards

**Comment by le-patenteux** [comment]
For anyone looking, I was able to migrate my DB from an external tool... It would not work for me using `occ db:convert-type` and any fix proposed earlier...

I reverted to using `pgloader` and so far, my database seems intact and everything is working fine!...
Here is a sanitized version of my bash command:

`pgloader mysql://mariadb_user:mariadb_password@192.168.0.23:3306/mariadb_database postgresql://pgsql_user:pgsql_password@192.168.0.23:5432/pgsql_database`

replace with your own parameters...

**Comment by le-patenteux** [comment]
I corrected my last comment... It broke!

Please, Nextcloud team, fix the DB migration tool! My MariaDB instance is not as stable as I would like it to be, an now, I can't migrate my database to postgreSQL.

I lack the technical knowlege to do it myself, and the proposed manual fixes listed earlier do not work for me (it fails to create the db shema)

**Comment by UltimateByte** [comment]
I hate to bump, but it's been 7 months and it's almost Christmas. A fix as a Christmas gift (even a late one) would be lovely.
As a (small) web hosting provider, I still need to convert one instance that I host, from PostgreSQL to MariaDB, in order to have it be ISO with the rest of my instances.

**Comment by xalt7x** [comment]
[PR 48666](https://github.com/nextcloud/server/pull/48666) to re-enable DB conversion is still active.
Perhaps, people can share there their experience with patches from PR.

**Comment by UltimateByte** [comment]
For anyone wondering, Nextcloud 31 should include this fix. According to the NC 31's Milestone, it should be released on Feb 25.
It was also backported to NC 30 and 29, so these should have patches for these versions as well.

Many thanks to everyone who worked on this!

**Comment by mamoit** [comment]
Do you know if the fix was included in `30.0.6` (released on the 12/02/2025)?
I don't see it in the changelog.

**Comment by joshtrichards** [comment]
Backport merged for both v30 and v29:

#50563
#59464

**Comment by rohlandm** [comment]
I can confirm, migrating my instance from MariaDB to Postgres worked without any issue 👌

**Comment by le-patenteux** [comment]
I get 3 duplicate key errors when trying to convert the database...

` An exception occurred while executing a query: SQLSTATE[23505]: Unique violation: 7 ERROR:  duplicate key value violates unique constraint "oc_collectives_pkey"  
  DETAIL:  Key (id)=(1) already exists. `

And no amount of googling found an answer on how to fix this!

**Comment by Tailerusin** [comment]
![Image](https://github.com/user-attachments/assets/7c6cba09-e669-46d7-a922-2ea46570b194)

**Comment by Tailerusin** [comment]
And yet it does not solve the problem of migration from mysql to postgresql.
I tried to migrate, but I get this error.

![Image](https://github.com/user-attachments/assets/0a7b24ad-78c1-4aee-bdaa-df5225b604cb)

**Comment by provokateurin** [comment]
Please open new issues instead of bumping this issue which is about a related but different topic.
